### PR TITLE
Prune index source columns

### PR DESCRIPTION
**Pull request by alandpost**
Add PruneIndexSourceColumns and its associated tooling:
* a builder function for IndexSourceNode
* a matcher for IndexSourceNode
* a bulider function for the matcher
* the rule itself
* a unit test for the rule
* enabling the rule in presto's rule list

Also, refactor some TableScanNode matching code so it can be shared
between TableScanNode and IndexSourceNode, because they are similar.

https://github.com/prestodb/presto/issues/7154

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java [hunk 1]
```diff
@@ -64,35 +63,8 @@ public MatchResult detailMatches(PlanNode node, PlanNodeCost cost, Session sessi
...
-            if (!actualConstraint.getDomains().get().containsKey(columnHandle)) {
-                return false;
-            }
-            if (!expectedColumnConstraint.getValue().contains(actualConstraint.getDomains().get().get(columnHandle))) {
-                return false;
-            }
-        }
-
-        return true;
+                        ((!expectedConstraint.isPresent()) ||
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java]: This should be in separate commit. It has nothing to do with `IndexSource` pruning.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java [hunk 2]
```diff
@@ -64,35 +63,8 @@ public MatchResult detailMatches(PlanNode node, PlanNodeCost cost, Session sessi
         String actualTableName = tableMetadata.getTable().getTableName();
         return new MatchResult(
                 expectedTableName.equalsIgnoreCase(actualTableName) &&
-                        domainMatches(tableScanNode, session, metadata));
-    }
-
-    private boolean domainMatches(TableScanNode tableScanNode, Session session, Metadata metadata)
-    {
-        if (!expectedConstraint.isPresent()) {
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java]: Why you decided to extract this outside of `domainMatches` even though in both cases where you use it, you check it anyway?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneIndexSourceColumns.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
+    {
+        closeAllRuntimeException(tester);
+        tester = null;
+    }
+
+    private PlanNode buildProjectedIndexSource(PlanBuilder p, Predicate<Symbol> projectionFilter)
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneIndexSourceColumns.java]: `static`

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneIndexSourceColumns.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    @Test
+    public void testAllOutputsReferenced()
+    {
+        tester.assertThat(new PruneIndexSourceColumns())
+                .on(p -> buildProjectedIndexSource(p, symbol -> true))
+                .doesNotFire();
+    }
+}
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneIndexSourceColumns.java]: Add test for no symbols referenced.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/IndexSourceNode.java [hunk 5]
```diff
@@ -67,6 +67,11 @@ public IndexSourceNode(
         checkArgument(!outputSymbols.isEmpty(), "outputSymbols is empty");
         checkArgument(assignments.keySet().containsAll(lookupSymbols), "Assignments do not include all lookup symbols");
         checkArgument(outputSymbols.containsAll(lookupSymbols), "Lookup symbols need to be part of the output symbols");
+        Set<ColumnHandle> assignedColumnHandles = ImmutableSet.copyOf(assignments.values());
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/IndexSourceNode.java]: I had to go look up how to do this.  This page seemed to be a good resource:  https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java (hunk 1 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java]: If it was in a separate commit, the factored-out function would have only one caller.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java (hunk 2 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java]: It's a question of the meaning of the optional.  The factored out function now verifies that the optional item exists, or not, as specified.  The fact that each matcher uses an empty optional to indicate "don't care" is separate.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneIndexSourceColumns.java (hunk 4 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneIndexSourceColumns.java]: In that case, the rewrite drops the lookup key, and fails an assertion in the IndexSourceNode constructor.  I copied the logic faithfully from PruneUnreferencedOutputs -- it just didn't have any testing, so of course it didn't test this.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ColumnReference.java [hunk 6]
```diff
@@ -44,30 +45,42 @@ public ColumnReference(String tableName, String columnName)
...
+        Optional<ColumnHandle> columnHandle = getColumnHandle(tableHandle, session, metadata);
 
         checkState(columnHandle.isPresent(), format("Table %s doesn't have column %s. Typo in test?", tableName, columnName));
 
-        return getAssignedSymbol(tableScanNode, columnHandle.get());
+        return getAssignedSymbol(assignments, columnHandle.get());
     }
 
-    private Optional<Symbol> getAssignedSymbol(TableScanNode tableScanNode, ColumnHandle columnHandle)
+    private Optional<Symbol> getAssignedSymbol(Map<Symbol, ColumnHandle> assignments, ColumnHandle columnHandle)
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ColumnReference.java]: `static`

Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ColumnReference.java]: Can't do it, because the checkState error message uses the table name and column name members.

Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ColumnReference.java]: The conflict looks to me to have been resolved, but the github UI says it's still there.

Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ColumnReference.java]: I think the problem was that I was doing git fetch and not git fetch --all.  Resolved now.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/IndexSourceNode.java [hunk 7]
```diff
@@ -67,6 +67,11 @@ public IndexSourceNode(
...
+        effectiveTupleDomain.getDomains().ifPresent(handleToDomain ->
+                checkArgument(
+                        assignedColumnHandles.containsAll(handleToDomain.keySet()),
+                        "Tuple domain handles must have assigned symbols"));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/IndexSourceNode.java]: can you extract the code you added as a method, so it will be easier to read this as you give a method name which would work a comment describing what it does

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestingConnectorIndexHandle.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner;
+
+import com.facebook.presto.spi.ConnectorIndexHandle;
+import com.fasterxml.jackson.annotation.JsonValue;
+
+public class TestingConnectorIndexHandle
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestingConnectorIndexHandle.java]: I think the convention is to create an enum for such empty implementations

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestingConnectorTransactionHandle.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner;
+
+import com.facebook.presto.spi.connector.ConnectorTransactionHandle;
+import com.fasterxml.jackson.annotation.JsonValue;
+
+public class TestingConnectorTransactionHandle
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestingConnectorTransactionHandle.java]: use enum here

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/IndexSourceNode.java (hunk 7 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/IndexSourceNode.java]: I was thinking that the checkArgument error string did the description for the reader.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/IndexSourceMatcher.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        checkState(shapeMatches(node), "Plan testing framework error: shapeMatches returned false in detailMatches in %s", this.getClass().getName());
+
+        IndexSourceNode indexSourceNode = (IndexSourceNode) node;
+        TableMetadata tableMetadata = metadata.getTableMetadata(session, indexSourceNode.getTableHandle());
+        String actualTableName = tableMetadata.getTable().getTableName();
+        return new MatchResult(
+                expectedTableName.equalsIgnoreCase(actualTableName) &&
+                        ((!expectedConstraint.isPresent()) ||
+                                domainsMatch(expectedConstraint, indexSourceNode.getEffectiveTupleDomain(), indexSourceNode.getTableHandle(), session, metadata)));
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/IndexSourceMatcher.java]: separate parameters with new lines

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/IndexSourceMatcher.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    public MatchResult detailMatches(PlanNode node, PlanNodeCost cost, Session session, Metadata metadata, SymbolAliases symbolAliases)
+    {
+        checkState(shapeMatches(node), "Plan testing framework error: shapeMatches returned false in detailMatches in %s", this.getClass().getName());
+
+        IndexSourceNode indexSourceNode = (IndexSourceNode) node;
+        TableMetadata tableMetadata = metadata.getTableMetadata(session, indexSourceNode.getTableHandle());
+        String actualTableName = tableMetadata.getTable().getTableName();
+        return new MatchResult(
+                expectedTableName.equalsIgnoreCase(actualTableName) &&
+                        ((!expectedConstraint.isPresent()) ||
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/IndexSourceMatcher.java]: I would extract this condition as variable and do cascade of `if`s

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java (hunk 1 above)
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java]: one caller is nothing bad

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/Util.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    static boolean domainsMatch(
+            Optional<Map<String, Domain>> expectedDomains,
+            TupleDomain<ColumnHandle> actualConstraint,
+            TableHandle tableHandle,
+            Session session,
+            Metadata metadata)
+    {
+        Optional<Map<ColumnHandle, Domain>> actualDomains = actualConstraint.getDomains();
+
+        if (expectedDomains.isPresent() != actualDomains.isPresent()) {
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/Util.java]: do we care `actualDomains` if expected is not set?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java [hunk 13]
```diff
@@ -64,35 +63,8 @@ public MatchResult detailMatches(PlanNode node, PlanNodeCost cost, Session sessi
...
-                return false;
-            }
-            if (!expectedColumnConstraint.getValue().contains(actualConstraint.getDomains().get().get(columnHandle))) {
-                return false;
-            }
-        }
-
-        return true;
+                        ((!expectedConstraint.isPresent()) ||
+                                domainsMatch(expectedConstraint, tableScanNode.getCurrentConstraint(), tableScanNode.getTable(), session, metadata)));
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java]: extract `expectedConstraint` (call `get()`)

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java (hunk 1 above)
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java]: alternatively you can check this optional in the called method body. That would even better as more would handled by the called method and not the caller code.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneIndexSourceColumns.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+        extends BaseRuleTest
+{
+    @Test
+    public void testNotAllOutputsReferenced()
+    {
+        tester().assertThat(new PruneIndexSourceColumns())
+                .on(p -> buildProjectedIndexSource(p, symbol -> symbol.getName().equals("orderkey")))
+                .matches(
+                        strictProject(
+                                ImmutableMap.of("x", PlanMatchPattern.expression("orderkey")),
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneIndexSourceColumns.java]: static imort?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneIndexSourceColumns.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+        ProjectNode parent = (ProjectNode) node;
+
+        PlanNode source = lookup.resolve(parent.getSource());
+        if (!(source instanceof IndexSourceNode)) {
+            return Optional.empty();
+        }
+
+        IndexSourceNode indexSourceNode = (IndexSourceNode) source;
+
+        Optional<Set<Symbol>> prunedOutputs = pruneInputs(indexSourceNode.getOutputSymbols(), parent.getAssignments().getExpressions());
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneIndexSourceColumns.java]: OK

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneIndexSourceColumns.java [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+        if (!prunedOutputs.isPresent()) {
+            return Optional.empty();
+        }
+
+        Set<Symbol> newLookupSymbols = indexSourceNode.getLookupSymbols().stream()
+                .filter(prunedOutputs.get()::contains)
+                .collect(toImmutableSet());
+
+        Map<Symbol, ColumnHandle> newAssignments = Maps.filterEntries(
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneIndexSourceColumns.java]: I can't use filterKeys, at the expression uses entry.getValue

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneIndexSourceColumns.java [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+                indexSourceNode.getAssignments(),
+                entry -> prunedOutputs.get().contains(entry.getKey()) ||
+                        indexSourceNode.getEffectiveTupleDomain().getDomains()
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneIndexSourceColumns.java]: OK

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/IndexSourceMatcher.java (hunk 10 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/IndexSourceMatcher.java]: OK

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/IndexSourceMatcher.java (hunk 11 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/IndexSourceMatcher.java]: I rearranged it, hopefully in a way that you like better.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java (hunk 13 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java]: The confusing thing here is the meaning of Optional<Foo> in the matcher, given a node field of type Optional<Foo>.  In the case of an empty option in the matcher, does that mean don't-care or must-be-empty?  The previous TableScan matcher code considered it to be don't-care, and I've preserved that behavior in botch matchers.  The underlying domainsMatch() function interprets an empty option as must-be-empty.  I'll add a comment to domainsMatch().  In the longer term, the builder approach I suggested for WindowMatcher addresses this confusion nicely IMO.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/Util.java (hunk 12 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/Util.java]: See above.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneIndexSourceColumns.java (hunk 14 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneIndexSourceColumns.java]: OK.  The reason this has been coming up is that in some tests we use both PlanBuilder.expression() and PlanMatchPattern.expression(), and I was importing the former.  This test, as you point out, only uses PlanMatchPattern.expression(), and so can import it.  I'll make the change.

**Comment by alandpost** [comment]
@kokosing Please take another look.  Thanks.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/Util.java [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+        Map<String, ColumnHandle> columnHandles = metadata.getColumnHandles(session, tableHandle);
+        for (Map.Entry<String, Domain> expectedColumnConstraint : expectedDomains.get().entrySet()) {
+            if (!columnHandles.containsKey(expectedColumnConstraint.getKey())) {
+                return false;
+            }
+            ColumnHandle columnHandle = columnHandles.get(expectedColumnConstraint.getKey());
+            if (!actualDomains.get().containsKey(columnHandle)) {
+                return false;
+            }
+            if (!expectedColumnConstraint.getValue().contains(actualDomains.get().get(columnHandle))) {
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/Util.java]: why `contains` and not `equals`?

Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/Util.java]: I don't know.  My intent was to move the logic unchanged from TableScanMatcher:  https://github.com/prestodb/presto/blob/77de222ade1b1ddfc207534663224475b8979131/presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java#L98
